### ISSUE TITLE
Compiler crash from composing an index and a get on the same line

### ISSUE THREAD

**Issue by duvenaud** [google-research/dex-lang#245]
I can write a line of code that gives a compiler bug, but when I break it into two lines it works.

Here's my mostly minimal working example.  The following code compiles fine:

```
def UpperTriMat (n:Type) : Type = i:n=>(i..)=>Float

def upperTriIndex (ref:Ref h (i':n=>(i'..)=>Float)) (i:n) : Ref h ((i..)=>Float) =
                       %indexRef ref i

def ok (_:Eq n) ?=> (a:n) : Unit =
  init_upper = for i:n. for j:(i..). 0.0

  withState init_upper \uRef.
    for i:n.
      a = upperTriIndex uRef ((ordinal i)@_)
      b =  get a
      ()
    for i:n. ()
  ()
```

But if I combine the two inner lines into one, I get a compiler bug:

```
def notok (_:Eq n) ?=> (a:n) : Unit =
  init_upper = for i:n. for j:(i..). 0.0

  withState init_upper \uRef.
    for i:n.
      b2 = get (upperTriIndex uRef ((ordinal i)@_))
      ()
    for i:n. ()
  ()
Compiler bug!
Please report this at github.com/google-research/dex-lang/issues

Non-global free variable in module: tmp13
.
.
.
```

**Comment by apaszke** [comment]
A simpler repro:
```
def idx (arr: i':n=>(i'..)=>Float) (i:n) : ((i..)=>Float) = arr.i
def eid (a:Type) -> (x:a) : a = x

def notok (_:Eq n) ?=> (a:n) : Unit =
  init_upper = for i:n. for j:(i..). 0.0
  -- XXX: This cannot be reduced
  -- eid (((0@n)..)=>Float) (idx init_upper (0@n))
  -- This should not replicate the term, but use the result of 0@n on the right!
  eid _ (idx init_upper (0@n))
  ()
```
I think (I haven't verified this yet) that the reason for the failure is that the synthesis pass replaces the `_` argument to `eid` with `((0@n)..)=>Float`, which leads to two problems:
1. I think that the synthesis pass doesn't expect the expressions it generates to emit any decls, so it just drops them, leading to the error message you're seeing.
2. Even if we emit the code, it will still lead to a type error, because the result of evaluating `0@n` on the left will not be bound to the same variable as the result of evaluating the `0@n` argument to `idx`.

**tl;dr** I think that the synthesis pass is broken when it emits non-atomic expressions. The program in the original description compiles just fine if `((ordinal i)@_)` is replaced with `i`, because then the inferred term is an atom.
